File: packages/android/locanara/src/main/kotlin/com/locanara/builtin/TranslateChain.kt
package com.locanara.builtin

import com.locanara.TranslateResult
import com.locanara.composable.Chain
import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import com.locanara.core.GenerationConfig
import com.locanara.core.LocanaraDefaults
import com.locanara.core.LocanaraModel

/**
 * Built-in chain for language translation.
 */
class TranslateChain(
    private val model: LocanaraModel = LocanaraDefaults.model,
    private val sourceLanguage: String = "en",
    private val targetLanguage: String
) : Chain {
    override val name = "TranslateChain"

    override suspend fun invoke(input: ChainInput): ChainOutput {
        val sourceLang = java.util.Locale.forLanguageTag(sourceLanguage)
            .getDisplayLanguage(java.util.Locale.ENGLISH).ifEmpty { sourceLanguage }
        val targetLang = java.util.Locale.forLanguageTag(targetLanguage)
            .getDisplayLanguage(java.util.Locale.ENGLISH).ifEmpty { targetLanguage }

        val prompt = BuiltInPrompts.translate.format(
            mapOf("text" to input.text, "sourceLang" to sourceLang, "targetLang" to targetLang)
        )

        println("[TranslateChain] input: ${input.text.take(200)}")
        val response = model.generate(prompt, GenerationConfig.STRUCTURED)
        val translated = BuiltInPrompts.stripPreamble(response.text.trim())
        println("[TranslateChain] output: $translated")

        val result = TranslateResult(
            translatedText = translated,
            sourceLanguage = sourceLanguage,
            targetLanguage = targetLanguage,
            confidence = 0.90
        )

        return ChainOutput(
            value = result,
            text = translated,
            metadata = input.metadata,
            processingTimeMs = response.processingTimeMs
        )
    }

    /** Type-safe execution that returns [TranslateResult] directly. */
    suspend fun run(text: String): TranslateResult {
        val output = invoke(ChainInput(text = text))
        return output.typed<TranslateResult>()
            ?: throw IllegalStateException("Unexpected output type from TranslateChain")
    }
}

File: packages/apple/Sources/BuiltIn/TranslateChain.swift
import Foundation

/// Built-in chain for language translation.
///
/// ```swift
/// let result = try await TranslateChain(targetLanguage: "ko").run("Hello world")
/// ```
@available(iOS 15.0, macOS 14.0, *)
public struct TranslateChain: Chain {
    public let name = "TranslateChain"

    private let model: any LocanaraModel
    private let sourceLanguage: String
    private let targetLanguage: String

    public init(
        model: (any LocanaraModel)? = nil,
        sourceLanguage: String = "en",
        targetLanguage: String
    ) {
        self.model = model ?? LocanaraDefaults.model
        self.sourceLanguage = sourceLanguage
        self.targetLanguage = targetLanguage
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        let sourceLang = Locale(identifier: "en")
            .localizedString(forIdentifier: sourceLanguage) ?? sourceLanguage
        let targetLang = Locale(identifier: "en")
            .localizedString(forIdentifier: targetLanguage) ?? targetLanguage

        let prompt = try BuiltInPrompts.translate.format([
            "text": input.text,
            "sourceLang": sourceLang,
            "targetLang": targetLang
        ])

        print("[TranslateChain] input: \(input.text.prefix(200))")
        let response = try await model.generate(prompt: prompt, config: .structured)
        let translated = BuiltInPrompts.stripPreamble(response.text.trimmingCharacters(in: .whitespacesAndNewlines))
        print("[TranslateChain] output: \(translated)")

        let result = TranslateResult(
            translatedText: translated,
            sourceLanguage: sourceLanguage,
            targetLanguage: targetLanguage,
            confidence: 0.90
        )

        return ChainOutput(
            value: result,
            text: translated,
            metadata: input.metadata,
            processingTimeMs: response.processingTimeMs
        )
    }

    /// Type-safe execution that returns `TranslateResult` directly.
    public func run(_ text: String) async throws -> TranslateResult {
        let output = try await invoke(ChainInput(text: text))
        guard let result = output.typed(TranslateResult.self) else {
            throw LocanaraError.executionFailed("Unexpected output type from TranslateChain")
        }
        return result
    }
}
